source/change-streams/change-streams.rst

```diff
@@ -169,6 +208,12 @@ Driver API
      */
     private resumeToken: Document;
 
+    /**
+     * The most recent promisedMinResumeToken returned in an aggregate or
+     * getMore response. For pre-4.2 versions of MongoDB, this remains unset.
+     */
+    private minPromisedResumeToken: Document;
```

Comment:
Reading the description of [SPEC-1122](https://jira.mongodb.org/browse/SPEC-1122) again, I agree that drivers need only track a single resume token.

I would think we could use the existing `resumeToken` property and simply change its documentation. So instead of:

```
  /**
   * The resume token (_id) of the document the iterator last returned
   */
  private resumeToken: Document;

  /**
   * The most recent promisedMinResumeToken returned in an aggregate or
   * getMore response. For pre-4.2 versions of MongoDB, this remains unset.
   */
  private minPromisedResumeToken: Document;
```

We have:

```
  /**
   * The resume token. This will either be the (_id) of the document most recently
   * returned by the iterator, or the most recent promisedMinResumeToken returned
   * in an aggregate or getMore response (if available).
   */
  private resumeToken: Document;
```

This would also eliminate other language in the current PR:

> When the change stream has returned all documents in a (possibly empty) batch (i.e. it needs to send a ``getMore``), it MUST overwrite the cached ``resumeToken`` with the value of the cached ``minPromisedResumeToken``.

Reply:
Yes, only one resume token needs to be tracked at a given time.

I'm thinking something closer to:
```
/**
   * The resume token. This will either be the (_id) of the document most recently
   * returned by the iterator, or the most recent promisedMinResumeToken returned
   * in an aggregate or getMore response (if available, and all documents in the most
   * recent batch have been returned).
   */
  private resumeToken: Document;
```

I want to clearly convey that you should only set resumeToken to minPromisedResumeToken after all documents in the current batch have been returned.

Reply:
> Yes, only one resume token needs to be tracked at a given time.

This works when the application is guaranteed to process all documents returned in a batch, but does not work for the case when an application is processing documents individually and may stop at any time. In the latter case resume token for each document still must be tracked to resume from that document onward.

Reply:
> To clarify: are we guaranteed to receive back a minPromisedResumeToken on an aggregate call, and will that be useable? Would this eliminate the need for startAtOperationTime?

@daprahamian: In https://github.com/mongodb/specifications/pull/413#discussion_r231524556, @cswanson310 is proposing have 4.2+ servers always return `minPromisedResumeToken` and potentially back-porting it to earlier versions if needed. If `minPromisedResumeToken` shows up in an `aggregate` or `getMore` response, that makes `startAtOperationTime` obsolete because we now have an actual resume token; however, drivers should maintain functionality for `startAtOperationTime` in the event we're talking to an older server (we really won't know until that first `aggregate` response comes in).

> all documents in the most recent batch have been returned

@kevinAlbs: I'm curious how this would actually come up. If we're iterating through a local batch of documents (from the initial `aggregate` or a previous `getMore`), the driver is capturing the `minPromisedResumeToken` on that response and then overwriting it as we observe `_id` values of documents within that batch (if any). If at any point down the line we issue a new `getMore` or `aggregate` command, it's either because we've reached the end of a batch and need more documents or we encountered a resumable error on that `getMore` and need to issue a new `aggregate` to reconstruct the cursor. In either event, there would be no more documents left to iterate in the last local batch.

@p-mongo: On the same token (no pun intended), I don't follow how your scenario is any different. Even if documents in the change stream are being processed individually, they are obtained from the change stream cursor's local batch in a particular order. Assuming there are documents in the batch, the resume token should always be the `_id` of the most recently returned document from the change stream (based on the iterator's current position).

Reply:
Good point. A change stream can't get a resumable error while returning documents in the middle of a batch and not doing any IO. However, if we provide a public change stream function like `getResumeToken`, then we need to keep track of what resume token to return.

Reply:
Agreed that this is necessary to support the getResumeToken() interface. Regarding the language, I'd propose the following:

`The resume token. In the event that the iterator has returned all documents in the most recent batch, this MUST be the 'minPromisedResumeToken' for that batch. In the case where we are still iterating the batch, this MUST be the (_id) of the document most recently returned by the iterator.`

In particular, the driver's resume token should be set to 'minPromisedResumeToken' as soon as the final document in the current batch is returned, rather than the resume token of that final document.